### PR TITLE
Decouple endpoint ID from addressing

### PR DESCRIPTION
**Pull request by tgraf**
This decouples the endpoint ID from the addressing model and requires plugins such as the CNI plugin to no longer be aware of the endpoint ID at all. Endpoints can be referred to by either container ID or ip addressing information. This allows use of 3rd party IPAM plugins and enables disablement of IPv6.

Depends on: #6079 

*API Consequences:*
* The endpoint ID parameters in PUT /endpoint/{id} becomes meaningless and will be allocated instead
* Clients that blindly assume that the endpoint ID is valid after the API will break. No such clients are known.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6080)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-missed-k8s

**Comment by tgraf** [comment]
test-me-please

File: pkg/endpoint/id/allocator.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	if !pool.Remove(idpool.ID(id)) {
+		return fmt.Errorf("endpoint ID %d is already in use", id)
+	}
+
+	return nil
+}
+
+// Release releases an endpoint ID that was previously allocated or reused
+func Release(id uint16) error {
+	if !pool.Insert(idpool.ID(id)) {
```
Comment (ianvernon) [inline pkg/endpoint/id/allocator.go:64]: Why are you checking the result of `pool.Insert` instead of `pool.Release`?

File: cilium-health/launch/endpoint.go [hunk 2]
```diff
@@ -25,11 +25,9 @@ import (
 	"time"
```
Comment (ianvernon) [inline cilium-health/launch/endpoint.go:25]: Typo in commit: decopuled --> decoupled

File: pkg/endpointmanager/manager.go [hunk 3]
```diff
@@ -163,6 +178,10 @@ func Remove(ep *endpoint.Endpoint) {
 	defer mutex.Unlock()
 	delete(endpoints, ep.ID)
 
+	if err := endpointid.Release(ep.ID); err != nil {
+		log.WithError(err).Warning("Unable to relese endpoint ID")
```
Comment (ianvernon) [inline pkg/endpointmanager/manager.go]: release

File: pkg/endpointmanager/manager.go [hunk 4]
```diff
@@ -171,10 +171,14 @@ func Remove(ep *endpoint.Endpoint) {
 		delete(endpointsAux, endpointid.NewID(endpointid.DockerEndpointPrefix, ep.DockerEndpointID))
 	}
 
-	if ep.IPv4.String() != "" {
+	if ep.IPv4.IsSet() {
 		delete(endpointsAux, endpointid.NewID(endpointid.IPv4Prefix, ep.IPv4.String()))
 	}
 
+	if ep.IPv4.IsSet() {
```
Comment (ianvernon) [inline pkg/endpointmanager/manager.go]: This should be `ep.IPv6.IsSet()`

File: pkg/endpoint/id/id.go [hunk 5]
```diff
@@ -71,31 +68,15 @@ func ParseCiliumID(id string) (int64, error) {
...
-
 // ValidateID parses specified id and returns normalized id as string.
 func ValidateID(id string) (PrefixType, string, error) {
-	prefix, _, err := ParseID(id)
-	if err != nil {
-		return "", "", err
+	prefix, id := SplitID(id)
+	switch prefix {
+	case CiliumLocalIdPrefix, CiliumGlobalIdPrefix, ContainerIdPrefix, DockerEndpointPrefix, ContainerNamePrefix:
+	case PodNamePrefix, IPv4Prefix, IPv6Prefix:
```
Comment (ianvernon) [inline pkg/endpoint/id/id.go]: can't this case collapse in the prior one?

File: daemon/k8s_watcher_test.go [hunk 6]
```diff
@@ -1156,8 +1167,11 @@ func (ds *DaemonSuite) Test_missingK8sNamespaceV1(c *C) {
 		{
 			name: "endpointmanager doesn't contain any endpoint that is part of that namespace. Should be no-op",
 			setupArgs: func() args {
+				endpointmanager.RemoveAll()
```
Comment (jrajahalme) [inline daemon/k8s_watcher_test.go:746]: Why was this not needed before? Why is this needed now?

File: pkg/endpoint/id/id.go [hunk 7]
```diff
@@ -45,15 +49,21 @@ func NewID(prefix PrefixType, id string) string {
...
+		return NewID(IPv4Prefix, ip.String())
+	}
+
+	return NewID(IPv6Prefix, ip.String())
+}
+
 // SplitID splits ID into prefix and id. No validation is performed on prefix.
 func SplitID(id string) (PrefixType, string) {
-	if s := strings.Split(id, ":"); len(s) == 2 {
+	if s := strings.SplitN(id, ":", 2); len(s) == 2 {
```
Comment (jrajahalme) [inline pkg/endpoint/id/id.go]: Could use `i := strings.IndexByte(id, ':'); i >= 0` instead, it does not need to allocate memory. Really matters only if we do this a lot.

Comment (jrajahalme) [reply pkg/endpoint/id/id.go]: return would then be `return id[:i], id[i+1:]`

**Review by jrajahalme — APPROVED** [review]
We still map an IP address to an ID for `endpointmanager.Lookup()`, how is this decoupling the endpoint ID from addressing?

Other than this, looks good.

File: pkg/endpointmanager/manager.go [hunk 8]
```diff
@@ -59,7 +59,20 @@ func init() {
...
 // Insert inserts the endpoint into the global maps.
-func Insert(ep *endpoint.Endpoint) {
+func Insert(ep *endpoint.Endpoint) error {
+	if ep.ID != 0 {
+		if err := endpointid.Reuse(ep.ID); err != nil {
+			return fmt.Errorf("unable to reuse endpoint ID: %s", err)
+		}
+	} else {
+		id := endpointid.Allocate()
+		if id == uint16(0) {
```
Comment (aanm) [inline pkg/endpointmanager/manager.go:79]: shouldn't it be `id == NoID`?

File: pkg/endpointmanager/manager.go (hunk 3 above)
Comment (aanm) [inline pkg/endpointmanager/manager.go]: When would this ever error out?

**Review by aanm — DISMISSED** [review]

**Comment by tgraf** [comment]
test-me-please

File: pkg/endpoint/id/allocator.go (hunk 1 above)
Comment (tgraf) [reply pkg/endpoint/id/allocator.go:64]: Releasing the ID requires to reinsert it into the pool

File: pkg/endpointmanager/manager.go (hunk 8 above)
Comment (tgraf) [reply pkg/endpointmanager/manager.go:79]: No, it returns an uint16

File: pkg/endpointmanager/manager.go (hunk 3 above)
Comment (tgraf) [reply pkg/endpointmanager/manager.go]: Not sure, but we should not have failure scenarios without a warning of some sort.

File: pkg/endpoint/id/id.go (hunk 7 above)
Comment (tgraf) [reply pkg/endpoint/id/id.go]: Can be done outside of this PR. Definitely not against it.

File: daemon/k8s_watcher_test.go (hunk 6 above)
Comment (tgraf) [reply daemon/k8s_watcher_test.go:746]: Not sure why it was not done before, it was not being cleaned up before and relied on the tests never inserting colliding IDs.

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-missed-k8s

**Review by ianvernon — APPROVED** [review]

**Review by nebril — APPROVED** [review]
